src/event.cpp

```diff
@@ -159,6 +161,39 @@ void process_collision_events()
   simulation::time_event_collision.stop();
 }
 
+int64_t process_refill_events(int64_t remaining_work, int64_t source_offset)
```

Comment:
I think that wouldn't be good, since we'd not be reviving from a secondary in that case. Considering that we have to deal with the LCG seeds and global particle id's in the particle rebirth but not in secondary revival, I think it makes sense to keep them separated.

Reply:
That's true, though could be addressed by having the particle ID's be grabbed atomically from some global counter and re-initialized using that (as the particle seed should get set based on the particle ID as I recall). It would be nice in some sense to just tack on a 2 or 3 line conditional to the existing particle revival logic rather than build out a whole extra queue + event process. The other upside is that we don't have to introduce an interval variable -- the particles would be revived just as frequently as secondaries are. The downside is that we pack more logic into a single kernel, and adjacent particles are unlikely to require complete revival at the same time, so it would make things necessarily branchier.

Reply:
I would take less divergent code with a few more lines than divergent code with less lines any day of the week. In fact, I plan to adopt this same approach for secondary revival eventually with a scan and separate kernel since it's a fairly infrequent operation that can cause divergence.

I'd definitely like to test the method you propose at some point in the future since it indeed would keep more particles in flight. If we find it performs better, it won't be hard to incorporate that addition. For now though, I'd like to keep this as-is as a reference point.

Reply:
Thinking on this some more, I think you're right that we should group these things. However, I think it should be a bit different. In the refill event, we should first check if the particle is dead or has any secondaries, and restart it appropriately either from the fission bank or its secondary bank. This of course means removing revive_from_secondary from the surface crossing and collision kernels. Then we can do exhaustive_find_cell on all the revived particles en masse. I think it makes sense, because this way we can remove the need to ever call exhaustive_find_cell ever again after the neighbor lists are established. That function takes a non-negligible amount of time, and it's likely worthwhile to pay attention to not letting lone threads try to run it. I imagine that much of the time in surface crossing or collisions, you'll get one thread on a warp calling revive_from_secondary which will make the others wait.

Reply:
Makes sense! As it is now we have a very coarse grained approach to events, so experimenting with splitting them up here and there to reduce branching I think could end up being really useful.